comp/networkpath/npscheduler/npschedulerimpl/npscheduler.go

```diff
@@ -7,31 +7,324 @@
+
+	// Deps
+	epForwarder  eventplatform.Forwarder
+	logger       log.Component
+	metricSender metricsender.MetricSender
+	statsdClient ddgostatsd.ClientInterface
+
+	// Counters
+	receivedPathtestCount    *atomic.Uint64
+	processedTracerouteCount *atomic.Uint64
+
+	// Pathtest store
+	pathtestStore       *pathteststore.PathtestStore
+	pathtestInputChan   chan *common.Pathtest
+	pathtestProcessChan chan *pathteststore.PathtestContext
+
+	// Scheduling related
+	running       bool
+	workers       int
+	stopChan      chan struct{}
+	flushLoopDone chan struct{}
+	runDone       chan struct{}
+	flushInterval time.Duration
+
+	// structures needed to ease mocking/testing
+	TimeNowFn func() time.Time
+	// TODO: instead of mocking traceroute via function replacement like this
+	//       we should ideally create a fake/mock traceroute instance that can be passed/injected in NpScheduler
+	runTraceroute func(cfg traceroute.Config) (payload.NetworkPath, error)
 }
 
-func (s *npSchedulerImpl) ScheduleConns(_ []*model.Connection) {
+func newNoopNpSchedulerImpl() *npSchedulerImpl {
+	return &npSchedulerImpl{enabled: false}
+}
+
+func newNpSchedulerImpl(epForwarder eventplatform.Forwarder, logger log.Component, sysprobeYamlConfig config.Reader) *npSchedulerImpl {
+	workers := sysprobeYamlConfig.GetInt("network_path.workers")
+	pathtestInputChanSize := sysprobeYamlConfig.GetInt("network_path.input_chan_size")
+	pathtestProcessChanSize := sysprobeYamlConfig.GetInt("network_path.process_chan_size")
+	pathtestTTL := sysprobeYamlConfig.GetDuration("network_path.pathtest_ttl")
+	pathtestInterval := sysprobeYamlConfig.GetDuration("network_path.pathtest_interval")
+	flushInterval := sysprobeYamlConfig.GetDuration("network_path.flush_interval")
+
+	logger.Infof("New NpScheduler (workers=%d input_chan_size=%d pathtest_ttl=%s pathtest_interval=%s)",
+		workers,
+		pathtestInputChanSize,
+		pathtestTTL.String(),
+		pathtestInterval.String())
+
+	return &npSchedulerImpl{
+		enabled:     true,
+		epForwarder: epForwarder,
+		logger:      logger,
+
+		pathtestStore:       pathteststore.NewPathtestStore(pathtestTTL, pathtestInterval, logger),
+		pathtestInputChan:   make(chan *common.Pathtest, pathtestInputChanSize),
+		pathtestProcessChan: make(chan *pathteststore.PathtestContext, pathtestProcessChanSize),
+		flushInterval:       flushInterval,
+		workers:             workers,
+
+		metricSender: metricsender.NewMetricSenderStatsd(statsd.Client),
+
+		receivedPathtestCount:    atomic.NewUint64(0),
+		processedTracerouteCount: atomic.NewUint64(0),
+		TimeNowFn:                time.Now,
+
+		stopChan:      make(chan struct{}),
+		runDone:       make(chan struct{}),
+		flushLoopDone: make(chan struct{}),
+
+		statsdClient: statsd.Client,
+
+		runTraceroute: runTraceroute,
+	}
+}
+
+func (s *npSchedulerImpl) ScheduleConns(conns []*model.Connection) {
 	if !s.enabled {
 		return
 	}
-	// TODO: IMPLEMENTATION IN SEPARATE PR (to make PRs easier to review)
+	startTime := time.Now()
```

Comment:
`s.TimeNowFn()`?

Reply:
Indeed, fixed and tested here 812d1ff8a7